### DISCUSSION TITLE
Adding local:ViewLocator in App.axaml

### DISCUSSION THREAD

**Discussion by lanibaldi-ors** [AvaloniaUI/Avalonia#11728] [General]
Hi, I'm using Avalonia (Version="0.10.21", Avalonia.Desktop" Version="0.10.21", ReactiveUI Version="0.10.21") on VS 2022.
Since this morning I'm trying to follow the tutorials - like ToDo List App - to implement a working sample of Model-View-ViewModel (MVVM) pattern. The problems come when I've reached the Locating Views step,  I add the ViewLocator specification in App.axaml:
```
<Application.DataTemplates>
        <local:ViewLocator/>
</Application.DataTemplates>
```
This is the message from compiler:
**Error		'local' is an undeclared prefix. Line 11, position 4.**		

Note: If I run the application without the specification above I get a window showing the name of the ViewModel

Any suggestion? Thanks, Luca

**Comment by kekekeks** [comment]
Add 
```cs


public class ViewLocator : IDataTemplate
{
    public IControl Build(object data)
    {
        if (data is null)
            return null;

        var name = data.GetType().FullName!.Replace("ViewModel", "View");
        var type = Type.GetType(name);

        if (type != null)
        {
            return (Control)Activator.CreateInstance(type)!;
        }

        return new TextBlock { Text = name };
    }

    public bool Match(object? data)
    {
        return data is ViewModelBase;
    }
}
```
to your project.

The locator was removed from 0.10.x templates by a mistake, I think.

**Reply by lanibaldi-ors** [reply]
Thanks, but it's what I did: it seems the ViewLocator is never hit.

**Reply by Mrxx99** [reply]
you need to add
`xmlns:local="using:NAMESPACE"` to the Application tag (where also `xmlns:x...` is) with NAMESPACE being the NAMESPACE of your ViewLocator

**Reply by lanibaldi-ors** [reply]
Yes, that's it. Now it likes the local prefix and then the window appears as expected.
Thanks a lot, have a nice weekend!

**Comment by timunie** [comment]
Let me add: please also check out Avalonia.Samples on GitHub. inside we explain the ViewLocator in detail.

**Reply by lanibaldi-ors** [reply]
I watched this sample below, but I've found the ViewLocator class and the local specification in the App.axaml. Am I wrong?

https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/BasicDataTemplateSample

**Reply by lanibaldi-ors** [reply]
I've checked the rest of the samples, it seems to me they are pretty much the same.
In particular:
https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample
